persistence/pre_persistence/app.go

```diff
@@ -0,0 +1,311 @@
+	}
+	maxRelaysToAddI, err := StringToBigInt(maxRelaysToAdd)
+	if err != nil {
+		return err
+	}
+	maxRelays.Add(maxRelays, maxRelaysToAddI)
+	// update values
+	app.MaxRelays = BigIntToString(maxRelays)
+	app.StakedTokens = BigIntToString(stakedTokens)
+	app.Chains = chainsToUpdate
+	// marshal
+	bz, err := cdc.Marshal(app)
+	if err != nil {
+		return err
+	}
+	return db.Put(key, bz)
+}
+
+func (m *PrePersistenceContext) DeleteApplication(address []byte) error {
+	if exists, _ := m.GetAppExists(address); !exists {
+		return fmt.Errorf("does not exist in world state")
+	}
+	db := m.Store()
+	key := append(AppPrefixKey, address...)
+	return db.Put(key, DeletedPrefixKey)
+}
+
+func (m *PrePersistenceContext) GetAppsReadyToUnstake(height int64, status int) (apps []*types.UnstakingActor, err error) { // TODO delete unstaking
+	db := m.Store()
+	unstakingKey := append(UnstakingAppPrefixKey, []byte(fmt.Sprintf("%d", height))...)
+	if has := db.Contains(unstakingKey); !has {
+		return nil, nil
+	}
+	val, err := db.Get(unstakingKey)
+	if err != nil {
+		return nil, err
+	}
+	if val == nil {
+		return make([]*types.UnstakingActor, 0), nil
+	}
+	unstakingApps := types.UnstakingActors{}
+	if err := proto.Unmarshal(val, &unstakingApps); err != nil {
+		return nil, err
+	}
+	for _, app := range unstakingApps.UnstakingActors {
+		apps = append(apps, app)
+	}
+	return
+}
+
+func (m *PrePersistenceContext) GetAppStatus(address []byte) (status int, err error) {
+	app, exists, err := m.GetApp(address)
+	if err != nil {
+		return ZeroInt, err
+	}
+	if !exists {
+		return ZeroInt, fmt.Errorf("does not exist in world state")
+	}
+	return int(app.Status), nil
+}
+
+func (m *PrePersistenceContext) SetAppUnstakingHeightAndStatus(address []byte, unstakingHeight int64, status int) error {
+	app, exists, err := m.GetApp(address)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return fmt.Errorf("does not exist in world state")
+	}
+	cdc := Cdc()
+	unstakingApps := types.UnstakingActors{}
+	db := m.Store()
+	key := append(AppPrefixKey, address...)
+	app.UnstakingHeight = unstakingHeight
+	app.Status = int32(status)
+	// marshal
+	bz, err := cdc.Marshal(app)
+	if err != nil {
+		return err
+	}
+	if err := db.Put(key, bz); err != nil {
+		return err
+	}
+	unstakingKey := append(UnstakingAppPrefixKey, []byte(fmt.Sprintf("%d", unstakingHeight))...)
+	if found := db.Contains(unstakingKey); found {
+		val, err := db.Get(unstakingKey)
+		if err != nil {
+			return err
+		}
+		if err := proto.Unmarshal(val, &unstakingApps); err != nil {
+			return err
+		}
+	}
+	unstakingApps.UnstakingActors = append(unstakingApps.UnstakingActors, &types.UnstakingActor{
+		Address:       app.Address,
+		StakeAmount:   app.StakedTokens,
+		OutputAddress: app.Output,
+	})
+	unstakingBz, err := cdc.Marshal(&unstakingApps)
+	if err != nil {
+		return err
+	}
+	return db.Put(unstakingKey, unstakingBz)
+}
+
+func (m *PrePersistenceContext) GetAppPauseHeightIfExists(address []byte) (int64, error) {
```

Comment:
NIT: Height should be `uint64`.

Might warrant a refactor TODO.

Reply:
This requires discussion. Tendermint uses int64

Reply:
Added to offline discussion list: https://www.notion.so/Andrew-Olshansky-1-1-d7d24cc5c6d54c3bad1f5c5af7b08b91#86be5aee6e0641029240125ceed7fbf3